patches/api/0387-More-Teleport-API.patch

```diff
@@ -35,51 +35,100 @@ index 0000000000000000000000000000000000000000..7713a5b7e06da185685f18e79eae2c97
++import org.jetbrains.annotations.ApiStatus;
 +
 +/**
-+ * Represents coordinates in a teleportation that should be handled relatively.
++ * Represents a flag that can be set on teleportation that may
++ * slightly modify the behavior.
 + *
-+ * @see org.bukkit.entity.Player#teleport(Location, PlayerTeleportEvent.TeleportCause, boolean, boolean, RelativeTeleportFlag...)
++ * @see EntityStateFlag
++ * @see RelativeTeleportFlag
 + */
-+@org.jetbrains.annotations.ApiStatus.Experimental
-+public enum RelativeTeleportFlag {
-+    /**
-+     * Represents the player's X coordinate
-+     */
-+    X,
-+    /**
-+     * Represents the player's Y coordinate
-+     */
-+    Y,
-+    /**
-+     * Represents the player's Z coordinate
-+     */
-+    Z,
++@ApiStatus.Experimental
++public sealed interface TeleportFlag permits TeleportFlag.EntityStateFlag, TeleportFlag.RelativeTeleportFlag {
++
 +    /**
-+     * Represents the player's yaw
++     * Note: These flags only work on {@link org.bukkit.entity.Player} entities.
++     * <p>
++     * Represents coordinates in a teleportation that should be handled relatively.
++     * <p>
++     * Coordinates of the location that the client should handle as relative teleportation
++     * Relative teleportation flags are only used client side, and cause the player to not lose velocity in that
++     * specific coordinate. The location of the teleportation will not change.
++     *
++     * @see org.bukkit.entity.Player#teleport(Location, PlayerTeleportEvent.TeleportCause, TeleportFlag...)
 +     */
-+    YAW,
++    @org.jetbrains.annotations.ApiStatus.Experimental
++    enum RelativeTeleportFlag implements TeleportFlag {
++        /**
++         * Represents the player's X coordinate
++         */
++        X,
++        /**
++         * Represents the player's Y coordinate
++         */
++        Y,
++        /**
++         * Represents the player's Z coordinate
++         */
++        Z,
++        /**
++         * Represents the player's yaw
++         */
++        YAW,
++        /**
++         * Represents the player's pitch
++         */
++        PITCH;
++    }
++
 +    /**
-+     * Represents the player's pitch
++     * Represents flags that effect the entity's state on
++     * teleportation.
 +     */
-+    PITCH;
++    @ApiStatus.Experimental
++    enum EntityStateFlag implements TeleportFlag {
++        /**
++         * If all passengers should not be required to be removed prior to teleportation.
++         *
++         * Note
++         * Teleporting to a different world with ignorePassengers to true while the entity has entities riding it
++         * will cause this teleportation to return false and not occur.
++         */
++        IGNORE_PASSENGERS,
++        /**
++         * If the entity should not be dismounted if they are riding another entity.
++         *
++         * Note:
++         * Teleporting to a different world with dismount to false while this entity is riding another entity will
```

Comment:
Same thing as above, "with dismount to false" sounds even more confusing, implying it can explicitly be set to false (?), "without this flag" or "when this flag is excluded" or however else